### ISSUE TITLE
current subtitle/audio streams not reported

### ISSUE THREAD

**Issue by jsgh** [xbmc/xbmc#16770]
I noticed that jsonrpc Player.GetProperties seems to always report currentsubtitle and currentaudiostream as the empty object ('{}'). After a bit of fiddling around I can tell that Player.SetAudioStream and Player.SetSubtitle do have the right effect: playback continues on with those settings.

And the current streams are *very briefly* reported correctly just after a change. However after a second they go back to blank.

Digging in it appears that retrieving the current streams ultimately ends up in CVideoPlayer::GetSubtitle(), which locks and reads from a local member. That member is updated by CVideoPlayer::UpdateContentState(), called from OpenStream(), which is called hardly ever. (And after adding a bit of debugging, seems to always set the three stream indices to -1 anyway.)

On the other hand, setting the current streams appears to go a completely different path once it hits CVideoPlayer, sending a message somewhere else to trigger the change. Ultimately the only state I can find which definitely does change are the AST_REG and SPST_REG DVD VM registers...

So within CVideoPlayer there appears to be no communication between the read side and the write side to synchronize the local state.

The reason why it *briefly* reports the correct stream after a change is back down in the GUI layer: when it sets a new stream it caches the value it is about to set locally for 1 second. During that second it won't call back into the player, just report its own value. As soon as the cache expires, it goes back to reporting the value it gets from the player, which is always -1.

I'm pretty sure this affects the OSD as well as jsonrpc, as I notice now that repeatedly clicking next audio/subtitle on the remote appears to cycle correctly through the available languages, except when the OSD flash first appears it appears to always say "Unknown". Then after leaving it for a bit, those buttons again show the current stream as "Unknown" before cycling through the valid stream names again.

**Comment by jsgh** [comment]
Ok, after a bit of digging around within CVideoPlayer:

m_SelectionStreams contains a list of streams with STREAM_SOURCE_NAV and demuxerId -1, with ids which are zero-based offsets of the stream within their type. (i.e. audio 0,1,2, subtitle 0,1,2).

But m_CurrentSubtitle/m_CurrentAudio contain streams with STREAM_SOURCE_DEMUX, a demuxerId > 0, and continuously increasing ids (i.e. audio 1,2,3, subtitle 4,5,6.)

So when UpdateContentState() tries to update m_content which is what the OSD and jsonrpc read from, it can't match the two different types of stream together, so sets the values in m_content to -1.

So, I think either one or the other of m_SelectionStreams and m_Current* are recording the wrong type of stream, or the match function TypeIndexOf() needs to be smarter so it can tell the streams are the same. I've no idea which.

A little help, please?

**Comment by jsgh** [comment]
(The id values from m_SelectionStreams appear to be what the lower levels expect to get given back to them: they likely couldn't intepret the stream id from m_Current* at all.)

**Comment by seldam** [comment]
I noticed that jsonrpc Player.GetProperties seems currentaudiostream only be filled when playing an mp4 for example. But when playing a movie from the movie library (kopied DVD) currentaudiostream is indeed empty.

**Comment by jsgh** [comment]
I can well believe that. I've done a bit more digging recently, and the situation is... complicated. Basically there are a bunch of *possible* identifiers for a particular subtitle stream when doing DVD, there isn't necessarily a one-to-one mapping between them all, and kodi fails to track them usefully enough when doing the switch.

I have a WIP patch that at least for me gets the right output, it also deals with something else I noticed, that kodi fails to distinguish between subtitles of the same language but different purpose (such as regular and large size subs) so for many DVDs you end up with like 3 or 4 subs all labelled "English" but no idea what the difference is. There are still cases where that happens (some DVDs just do that, mark several sub options *exactly* the same). I'm unconvinced by that part of the patch so far though so might end up changing it a bit more...

https://github.com/jsgh/xbmc/commit/762e9861cd321412a4975eccbd101982cfc273dc

**Comment by seldam** [comment]
Nice work @jsgh. I also noticed that currentvideostream is also an empty object or null. videostreams and audiostreams are beeing populated. It looks like currentaudiostream and currentvideostream share the same problem when playing a copied DVD from the movies library.

**Comment by jsgh** [comment]
Yes, and my change above also handles reporting currentaudiostream correctly.

currentvideostream is a bit weird though: DVD only allows for one video stream to be present at all - there is no possibility of having anything to select there, and the corresponding parts of the libdvdnav API provide no facilities or reporting related to that. So technically there's not much point in doing anything with it for the DVD case, though you could argue that the currentvideostream property ought to at least reflect that. (Other formats might contain multiple video streams of course, but not being played via libdvdnav may well already report it correctly... however not my use case.)

**Comment by malard** [comment]
I haven't looked too deeply, but is this still an open issue? if it is and someone confirms, I will take on this ticket

**Comment by jsgh** [comment]
Well, I'm still running with my local patch, which works well for me:

https://github.com/jsgh/xbmc/commit/cad2b034e3580b0e657b38d0d9add98f1ec79629
https://github.com/jsgh/xbmc/commits/jscombined-le10.0.3/

It doesn't look like much has changed on master so the issue is probably still there.

**Comment by malard** [comment]
do you wish to submit a PR i will review it and get it merged in? or tell me specifically what you changed that fixed it and I can support the process

Thanks

**Comment by jsgh** [comment]
Oof, I'm a bit busy atm. However these are the changes on that specific commit from my local set:

The changes are basically:
1. Add additional subtitle types used by DVD. Populate them from the DVD subpicture flags and export them over jsonrpc.
2. Record the DVD NAV id as an additional index/identifier for substreams. If we're playing DVD, then track changes coming from libdvdnav to that, and synchronise display state to it.
3. Add a couple of non-ISO (but non-clashing) language codes that are nevertheless used by DVDs.

**Comment by malard** [comment]
Okay, sounds like we can break this down into a few standalone PRs, the Language Codes looks like one, (I think we can safely add that without much regression testing)

do you have any example iso's the show this working/not working so i can test against?

**Comment by jsgh** [comment]
> do you have any example iso's the show this working/not working so i can test against?

Hmm. My collection is exact folder-structure copies of my physical DVD collection. (2560 discs total.) Not easy to share.

I can tell you that the two ISO 639-1 codes in question, "in" and "iw", are actually ex-standard codes that were deprecated in 1989-03-11:

https://www.loc.gov/standards/iso639-2/php/code_changes_bycode.php?code_ID=184
https://www.loc.gov/standards/iso639-2/php/code_changes_bycode.php?code_ID=207

They were formally changed to "id" and "he" respectively, for future versions of the standard.

However, changing a standard like that can have no effect on other standards that incorporate a specific version of them, nor can they generally cause a physical artefact produced according to an earlier version to become non-conformant. Such is the case for DVDs.

(I would add that if a particular form of non-conformance is widespread, then you should probably accommodate it anyway, but I have no need to push that line here since the two codes are previously standard.)

For "in", the only discs in my collection using lang "in" are The Raid (2011) and The Raid 2 (2014). Nothing else I have appears to have Indonesian audio or subs, under either the old code or the new code.

For "iw", this turns out to be a lot more common. 73 out of the above 2560 discs (mostly UK-bought region 2 discs). Here's a smaller selection:

Dr. Strangelove (1964)
MASH (1970)
Ghostbusters (1984)
Gattaca (1997) (2dvd widescreen+regular release, both discs have "iw" subs)
Constantine (2005)
I Am Legend (2007)
Skyfall (2012)
The Dark Tower (2017)

The consequences of not including these codes in the mappings is that when you click through the audio/subs they show up on the OSD as "Unknown", which irritates the tits off me to a disproportionate degree.

**Comment by malard** [comment]
*This reply was written and posted by Claude on @malard's behalf.*

Fixed on the `jsonrpc-14` integration branch (`927a76b8b4`, `7cb8f40283`): hiding a bitmap subtitle whose language differs from the audio track closes its stream, which dropped the selection permanently; an absent selection is also now `null` rather than `{}`.

That branch is the JSON-RPC v14 API landing as a single PR for v22, per the team's one-PR-per-topic convention. This report is resolved but stays open until the branch merges, so it is in a holding pattern rather than awaiting further investigation.
